linalg/sparsemat.cpp

```diff
@@ -2461,52 +2461,60 @@ void SparseMatrix::DiagScale(const Vector &b, Vector &x, double sc) const
    });
 }
 
-void SparseMatrix::Jacobi2(const Vector &b, const Vector &x0, Vector &x1,
-                           double sc) const
+template <bool useFabs>
+static void JacobiDispatch(const Vector &b, const Vector &x0, Vector &x1,
+                           const Memory<int> &I, const Memory<int> &J,
+                           const Memory<double> &A, const int height,
+                           const double sc)
 {
-   MFEM_VERIFY(Finalized(), "Matrix must be finalized.");
+   const bool useDevice = b.UseDevice() || x0.UseDevice() || x1.UseDevice();
 
-   for (int i = 0; i < height; i++)
+   const auto bp  = b.Read(useDevice);
+   const auto x0p = x0.Read(useDevice);
+   auto       x1p = x1.Write(useDevice);
+
+   const auto Ip = Read(I, height+1, useDevice);
+   const auto Jp = Read(J, J.Capacity(), useDevice);
+   const auto Ap = Read(A, J.Capacity(), useDevice);
+
+   MFEM_FORALL_SWITCH(useDevice, i, height,
    {
-      double resi = b(i), norm = 0.0;
-      for (int j = I[i]; j < I[i+1]; j++)
+      double resi = bp[i], norm = 0.0;
+      for (int j = Ip[i]; j < Ip[i+1]; j++)
       {
-         resi -= A[j] * x0(J[j]);
-         norm += fabs(A[j]);
+         resi -= Ap[j] * x0p[Jp[j]];
+         if (useFabs)
+         {
+            norm += fabs(Ap[j]);
+         }
+         else
+         {
+            norm += Ap[j];
+         }
       }
       if (norm > 0.0)
       {
-         x1(i) = x0(i) + sc * resi / norm;
+         x1p[i] = x0p[i] + sc * resi / norm;
       }
       else
       {
-         MFEM_ABORT("L1 norm of row " << i << " is zero.");
+         MFEM_ABORT_KERNEL("L1 norm of row is zero.");
```

Comment:
the printout here should be either L1 norm or the sum depending if the `fabs` is used or not. One other thing, we might want to allow negative sum? If not the printout should be modified accordingly.

Reply:
> the printout here should be either L1 norm or the sum depending if the fabs is used or not. 

Good point, fixed. 

> One other thing, we might want to allow negative sum? If not the printout should be modified accordingly.

Would we?

On a related tangent:
@tzanio @YohannDudouit (if I've missed someone else in mfem that is knowledgeable on this please tag them), I don't like the fact that we have to give up the row number in the error printout here but this is as a result of `MFEM_ABORT_KERNEL` being a `printf` in `iostream`'s clothing. Perfect world `MFEM_ABORT_KERNEL` would be have a `__VA_ARGS__` such that one could do an actual `printf`... or is there already such a macro?